slurm_ops_manager/nvidia.py

```diff
@@ -0,0 +1,167 @@
+                cmd = f"curl -s -L -o {path} {url}"
+                subprocess.run(shlex.split(cmd))
+
+                # add key
+                cmd = f"apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/{distro}/x86_64/7fa2af80.pub" # noqa
+                subprocess.run(shlex.split(cmd))
+
+                # it is necessary to update apt's db after every repo change
+                subprocess.call(["apt-get", "update"])
+            elif self._operating_system == "centos":
+                distro = "rhel7"
+                url = f"http://developer.download.nvidia.com/compute/cuda/repos/{distro}/x86_64/cuda-{distro}.repo" # noqa
+                cmd = f"yum-config-manager --add-repo {url}"
+                subprocess.run(shlex.split(cmd))
+
+                subprocess.call(["yum", "clean", "expire-cache"])
+
+        self._stored.gpu_repo_configured = True
+
+    @property
+    def package(self) -> str:
+        """Return nVdia package name."""
+        return self._stored.gpu_package_name
+
+    @package.setter
+    def package(self, pkg: str):
+        """Change the package to install."""
+        if not pkg:
+            # if empty, sets the default
+            if self._operating_system == "ubuntu":
+                self._stored.gpu_package_name = "cuda-drivers"
+            elif self._operating_system == "centos":
+                self._stored.gpu_package_name = "nvidia-driver-latest-dkms cuda cuda-drivers"
+        else:
+            self._stored.gpu_package_name = pkg
+
+    @property
+    def version(self):
+        """Return installed driver version."""
+        if self.installed:
+            pkg = self.package
+
+            if "ubuntu" == self._operating_system:
+                cmd = f"dpkg --status {pkg} | grep \"^Version\""
+                version = subprocess.check_output(cmd, shell=True)
+            elif "centos" == self._operating_system:
+                cmd = f"yum info -C {pkg} | grep \"^Version\""
+                locale = {"LC_ALL": "C", "LANG": "C.UTF-8"}
+                version = subprocess.check_output(cmd, shell=True, env=locale)
+            else:
+                logger.error("## Unsupported OS")
+                return ""
+
+            return version.decode().split(":")[-1].strip()
+        else:
+            return "not installed"
+
+    def install(self):
+        """Install Nvidia GPU packages.
+
+        If a custom repository was not previously configured, this method will
+        setup the default repository before installing the drivers.
+        """
+
+        if not self._stored.gpu_repo_configured:
+            self.repository = "" # set the default
+
+        logger.debug("#### Nvidia - detecting OS to install drivers")
+        uname = subprocess.check_output(shlex.split("uname -r")).strip().decode()
+
+        if self._operating_system == "ubuntu":
+            cmd = f"apt-get install --yes linux-headers-{uname} {self.package}"
```

Comment:
I wonder if it would be useful to create a command map for the OS? Something like this:

```python
cmd_map = dict(
    ubuntu=dict(
        install_drivers=f"yum install --assumeyes kernel-devel-{uname} kernel-headers-{uname} cuda-drivers",
    ),
    centos=dict(
        install_drivers=f"yum install --assumeyes kernel-devel-{uname} kernel-headers-{uname} nvidia-driver-latest-dkms cuda cuda-drivers",
    ),
)
```

with the other commands added as well.

Reply:
It took me a while to understand your snippet, but I think I got the idea. Will try it.